Sources/Amplitude/Amplitude.m

```diff
@@ -971,11 +984,11 @@ - (void)makeEventUploadPostRequest:(NSString *)url events:(NSString *)events num
                 if (numEvents == 1) {
                     if (maxEventId >= 0) {
                         [AMPStorage remove:[AMPStorage getDefaultEventsFile:self.instanceName]];
-                        self->_eventsBuffer = [[NSMutableArray alloc] init];
+                        self->_eventsBuffer = [self removeEventFromBuffer:self->_eventsBuffer currentEventString:events];
```

Comment:
More explanation here.
1. The backend will upload the events multiple times until there is no event.
2. if we received 413, we will update the `backoffUploadBatchSize` here (https://github.com/amplitude/Amplitude-iOS/pull/357/files#diff-8b27f2721223c88ca6e8487d0f8c71b7398c492580cc223f8b124672a7ac042fR1001) and try to upload a shorter array with events. 
With the original code, we have 2 events for example, if one of the events returns response 413 and the other one is good, we will lose the last events since we clean up the eventsBuffer after we try to send the first event. 
Please check test `testRequestTooLargeBackoffLogic`.